### PR TITLE
Add blog post for Structured Authz beta

### PR DESCRIPTION
**Pull request by ritazh**
Feature blog post for KEP 3221: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3221-structured-authorization-configuration
KEP issue: https://github.com/kubernetes/enhancements/issues/3221

### CODE REVIEW

**Comment by stealthybox** [comment]
I'm happy to help with this blog post -- we should include a kubeadm/kind example.
We may also want to update the [kubeadm apiserver flag docs](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#apiserver-flags) to show this can be used with `extraVolumes`.

**Comment by fkautz** [comment]
Hi @ritazh 👋🏽

I'm Frederick, a shadow from the Release Comms Team! #release-comms

I hope you're doing well! Just a friendly reminder that the deadline for the Kubernetes release blog post is coming up on March 22, 2024.  When you have a draft ready for review, please update the 'Blog Status' to 'Ready for Review'.

Thank you so much for your time and contribution!

**Comment by stealthybox** [comment]
Thanks for the reminder @fkautz !

@ritazh the main progress we've made on this is some quick usage examples of Structured Authz /w `kind`.
Lior also found and fixed a bug in kind related to structured authz.

I have not started on the copy to support the examples, but there is some good material in the use-cases from the KEP.

**Comment by sftim** [comment]
/hold
pending assignment of publication date

**Comment by ritazh** [comment]
@stealthybox feel free to push something if you have it. then I can integrate. I have given you push access.

**Comment by kcmartin** [comment]
@stealthybox  👋 Comms Team here, following up on your feature blog post for 1.30. The **Ready for Review** deadline is coming up this week on March 22, 2024. Please let us know if we can help out in any way!

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (palnabarun) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:5]: Thinking more about the title. Should we instead focus on "Authorization made more modular and powerful"?

The feature did start with the thought of adding multiple authz webhooks support but we achieved much more in the process.

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (palnabarun) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:161]:
```suggestion
startup has been a key demand. One of the blockers for having a controller
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (palnabarun) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:163]:
```suggestion
achieved through an authorization webhooks. This was not possible before as specifying multiple authorization webhooks in the handled chain was simply not possible.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (palnabarun) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:273]:
```suggestion
those will continue to work as-is. Specifying both `--authorization-config` and `--authorization-modes`/`--authorization-webhook-*` won't work. You need to drop the older flags from your kube-apiserver command.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:43]:
```suggestion
specifying multiple webhooks in the authorization chain (all other authorization types
are specified no more than once). Each webhook authorizer comes with
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:171]: this means "any service account in the kube-superuser namespace"... was that intentional? if not, maybe swap this out for a more clearly fake example group (maybe a non-serviceaccounts one)?

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:192]: CRDs are namespaced, so this condition means this webhook wouldn't intercept any CRD requests ... should this be 
```
- expression: request.resourceAttributes.resource.resource = "customresourcedefinitions"
- expression: request.resourceAttributes.resource.group = ""
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:196]:
```suggestion
      - expression: request.resourceAttributes.verb in ['update', 'patch', 'delete', 'deletecollection']
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:215]: copy any changes from the above CRD comments to this example

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:258]:
```suggestion
removed and the configuration file version will be promoted to v1.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (ritazh) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:5]: sounds good. 
how about `Enhancing Security: Simplified Multi-Webhook and Modular Authorization`
I would like to keep multi-webhook if possible since that is the main pain point and motivation driving this enhancement

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (ritazh) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:171]: I took this from the KEP. it's using 
> They can only be modified by users in the group "system:serviceaccounts:kube-superuser"

how about `admin` or `crd-admin`

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (ritazh) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:192]: Do you mean the CRDs (not CRs) are cluster scoped and therefore, we should remove this check 
>  expression: request.resourceAttributes.namespace == 'kube-system'

with

>  expression: request.resourceAttributes.resource.resource = "customresourcedefinitions" 
>  expression: request.resourceAttributes.resource.group = ""

Comment (liggitt) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:192]: oops, yes 🤦

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (liggitt) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:171]: yeah, either of those sgtm

**Review by palnabarun — APPROVED** [review]
/lgtm

**Comment by ritazh** [comment]
/assign @nate-double-u

Ready for review. Thanks!

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:163]:
```suggestion
startup has been a key demand. One of the blockers of having a controller
reconcile those CRDs is having a protection mechanism for them, which can be
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:166]:
```suggestion
authorization chain was simply not possible. Now, with the Structured
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:168]:
```suggestion
webhooks, offering a solution where RBAC falls short, especially when denying
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:156]:
```suggestion
The following configuration examples illustrate real-world scenarios that need
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:173]:
```suggestion
- They can only be modified by users in the group `admin`.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:204]:
```suggestion
handing them off to webhooks using frameworks like Open Policy Agent. In the
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:258]: > For Kubernetes v1.31, we expect the feature to stay in beta while we get more feedback.

I feel like this sentence is missing something -- is this only the process for v1.31? Is this is a general beta feature process? Who is providing the feedback?

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:260]:
```suggestion
feedback. Once it is ready for GA, the feature flag will be removed, and
the configuration file version will be promoted to v1.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:264]:
```suggestion
Kubernetes doc website. You can also follow along on with
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:265]:
```suggestion
[KEP-3221](https://kep.k8s.io/3221) to track progress in coming
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:262]:
```suggestion
Learn more about this feature on the [structured authorization
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:271]:
```suggestion
In this post, we have covered the benefits of the Structured Authorization
Configuration feature in Kubernetes v1.30 and a few sample configurations
for real-world scenarios. To use this feature, you must specify the path to the
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:310]:
```suggestion
implementors as they build their integrations with
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:317]:
```suggestion
If you are interested in helping develop this feature,
sharing feedback, or participating in any other ongoing SIG Auth projects, please
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:323]:
```suggestion
held every other Wednesday.
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:325]: US spelling:
```suggestion
## Acknowledgments
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (nate-double-u) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:328]:
```suggestion
This feature was driven by contributors from several different
companies. We would like to extend a huge thank you to everyone who contributed
their time and effort to make this possible.
```

**Review by nate-double-u — COMMENTED** [review]
A copy edit pass.

Since this is a release blog, I think that the release-comms folks probably have an opinion on the publication date. /cc @kcmartin

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (kcmartin) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:4]:
```suggestion
date: 2024-04-25
```
Please add the assigned publication date

Comment (sftim) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:4]: (this also means a file rename, I'm afraid)

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (sftim) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:13]:
```suggestion
```

or, as a subtitle (we don't usually have article subtitles)

```diff
-# Enhancing Kubernetes Authorization with Multiple Webhooks and Structured Configuration
+_Enhancing Kubernetes authorization with multiple webhooks and structured configuration_
```

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (sftim) [inline content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:3]:
```suggestion
title: 'Kubernetes 1.30: Multi-Webhook and Modular Authorization Made Much Easier'
```
?

I'm wary of calling it easy - our readers might disagree, even if the new thing really helps.

**Review by sftim — COMMENTED** [review]
Some more changes I recommend

File: content/en/blog/_posts/2024-04-xx-structured-authz-beta.md
Comment (ritazh) [reply content/en/blog/_posts/2024-04-xx-structured-authz-beta.md:258]: will update to
> From Kubernetes 1.30, the feature is in beta and enabled by default. For Kubernetes v1.31, we expect the feature to stay in beta while we get more feedback from users.

In general, we keep beta features around for one or two releases before promoting it to GA.

**Review by nate-double-u — COMMENTED** [review]
Thanks for making those updates! Docs lgtm.

/lgtm

**Comment by sftim** [comment]
Wait, how does this align with https://github.com/kubernetes/website/pull/45106 ?

I think we should put some work into telling readers how the two things are related.

**Comment by ritazh** [comment]
> Wait, how does this align with #45106 ?
> 
> I think we should put some work into telling readers how the two things are related.

@sftim How about if both blogs add something like the following:
> With CEL, administrators can craft layered security policies based on various factors, including user attributes, resource labels, and contextual information for both authorization and authentication. For more information on structured authentication configuration, refer to https://kubernetes.io/blog/2024/04/25/structured-authentication-moves-to-beta

I can add it after this: https://github.com/kubernetes/website/pull/45137/files#diff-ad10135436eec06f65b6ae8acd523b1123e01ad0cd3fa64c40292f830aac9ff6R49

File: content/en/blog/_posts/2024-04-25-structured-authz-beta.md
Comment (sftim) [reply content/en/blog/_posts/2024-04-25-structured-authz-beta.md:4]: (also, update the filename to match)

**Comment by sftim** [comment]
The other article now has this in its intro:
```markdown
Today's article is about _authentication_: finding out who's performing a task, and checking
that they are who they say they are. Check back in tomorrow to find about what's new in
Kubernetes v1.30 around _authorization_ (deciding what someone can and can't access).
```

My advice is, put it in early (before the first heading), and write it for an audience that is going to skip the blog article. It's tough to draft for the readers who don't care about this news, but it's helpful if we do serve them by making it obvious to someone that they're not going to need to read the rest of the article.

However, that's really soft advice. I expect we'll be fine with whatever wording you'd like to put @ritazh

**Comment by fkautz** [comment]
I agree with this comment above, I think it would be useful to state this is about **authorization** and that users interested in the complementary use case of authentication should also see the previous day's article.

@sftim  Would it make sense to perhaps make this explicit and say this is a two part blog post? Part 1: Authentication. Part 2: Authorization.

**Comment by fkautz** [comment]
@sftim the lgtm fell off from the last change. Just need the sign off if you think it looks good.

Thanks!

**Pull request merged**
